### PR TITLE
widgets: add TZ env var support for $TIME

### LINKED ISSUE #616: TZ (timezone) environment variable is being ignored
### Regression?

No

### Hyprlock Info and Version

Hyprlock version 0.5.0
<details> <summary>Hyprlock config</summary>

```sh
background {
  blur_passes=3
  blur_size=8
  path=screenshot
}

general {
  disable_loading_bar=true
  hide_cursor=true
  no_fade_in=false
}

label {
    monitor =
    text = $TIME
    color = rgba(200, 200, 200, 1.0)
    font_size = 200
    font_family = Noto Sans Extralight
    position = 0, 250
    halign = center
    valign = center
}

input-field {
  monitor=
  rounding = -1
  dots_center=true
  fade_on_empty=false
  font_color=rgb(202, 211, 245)
  inner_color=rgba(255, 255, 255, 0.1)
  outline_thickness=0
  placeholder_text=<i>Password of <b>$USER</b>?</i>
  position=0, 0
}
```
</details>

### Compositor Info and Version

<details> <summary>System/Version info</summary>

```sh
Hyprland 0.45.2 built from branch v0.45.2-b at commit 12f9a0d0b93f691d4d9923716557154d74777b0a  ([gha] Nix: update inputs).
Date: 2024-11-19
Tag: v0.45.2, commits: 12f9a0d0b93f691d4d9923716557154d74777b0a
built against aquamarine 0.4.4


flags set:
debug


System Information:
System name: Linux
Node name: monstrum
Release: 6.6.63
Version: #1-NixOS SMP PREEMPT_DYNAMIC Fri Nov 22 14:38:37 UTC 2024


GPU information:
03:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Oland GL [FirePro W2100] [1002:6608] (prog-if 00 [VGA controller])


os-release: ANSI_COLOR="1;34"
BUG_REPORT_URL="https://github.com/NixOS/nixpkgs/issues"
BUILD_ID="24.11.20241210.a0f3e10"
CPE_NAME="cpe:/o:nixos:nixos:24.11"
DEFAULT_HOSTNAME=nixos
DOCUMENTATION_URL="https://nixos.org/learn.html"
HOME_URL="https://nixos.org/"
ID=nixos
ID_LIKE=""
IMAGE_ID=""
IMAGE_VERSION=""
LOGO="nix-snowflake"
NAME=NixOS
PRETTY_NAME="NixOS 24.11 (Vicuna)"
SUPPORT_END="2025-06-30"
SUPPORT_URL="https://nixos.org/community.html"
VARIANT=""
VARIANT_ID=""
VENDOR_NAME=NixOS
VENDOR_URL="https://nixos.org/"
VERSION="24.11 (Vicuna)"
VERSION_CODENAME=vicuna
VERSION_ID="24.11"


plugins:


```

</details>

### Description

`hyprlock` is ignoring the `TZ` variable, which matters on systems with user in multiple timezones.

In my case it's a home server with UTC as the global timezone, but the media user sets `TZ=Europe/Berlin`.

For that user, `$TIME` is still in UTC, despite this user being in `Europe/Berlin`.

After I change the global `TZ` to `Europe/Berlin`, the `$TIME` is correct.

### How to reproduce

The time shown on the lock screen after launching it like:

```
TZ=Europe/Berlin hyprlock
```

should be different from:

```
TZ=UTC hyprlock
```

### Crash reports, logs, images, videos

_No response_

### PR DESCRIPTION
**Pull request by davc0n**
Closes #616

### CODE REVIEW

**Comment by davc0n** [comment]
Closes https://github.com/hyprwm/hyprlock/issues/616

File: src/renderer/widgets/IWidget.cpp [hunk 1]
```diff
@@ -120,43 +119,49 @@ static void replaceAllLayout(std::string& str) {
...
+    try {
+        auto name = std::getenv("TZ");
+        if (name) {
+            pCurrentTz = std::chrono::locate_zone(name);
+        }
+    } catch (std::runtime_error&) {
+        Debug::log(WARN, "Invalid TZ value. Falling back to current timezone!");
+    }
+
+    if (!pCurrentTz) {
```
Comment (vaxerski) [inline src/renderer/widgets/IWidget.cpp:133]: no {}

File: src/renderer/widgets/IWidget.cpp [hunk 2]
```diff
@@ -120,43 +119,49 @@ static void replaceAllLayout(std::string& str) {
...
 
 static bool        logMissingTzOnce = true;
-static std::string getTime() {
-    const auto PCURRENTTZ = std::chrono::current_zone();
-    const auto TPNOW      = std::chrono::system_clock::now();
+static std::chrono::hh_mm_ss<std::chrono::system_clock::duration> getTime() {
+    const std::chrono::time_zone* pCurrentTz = nullptr;
+    try {
+        auto name = std::getenv("TZ");
+        if (name) {
```
Comment (vaxerski) [inline src/renderer/widgets/IWidget.cpp:126]: no {}

File: src/renderer/widgets/IWidget.cpp [hunk 3]
```diff
@@ -120,43 +119,49 @@ static void replaceAllLayout(std::string& str) {
...
+    if (!pCurrentTz) {
         if (logMissingTzOnce) {
-            Debug::log(WARN, "Current timezone unknown for $TIME. Falling back to UTC!");
+            Debug::log(WARN, "Current timezone unknown. Falling back to UTC!");
             logMissingTzOnce = false;
         }
         hhmmss = std::chrono::hh_mm_ss{TPNOW - std::chrono::floor<std::chrono::days>(TPNOW)};
-    } else
-        hhmmss = std::chrono::hh_mm_ss{PCURRENTTZ->to_local(TPNOW) - std::chrono::floor<std::chrono::days>(PCURRENTTZ->to_local(TPNOW))};
+    } else {
```
Comment (vaxerski) [inline src/renderer/widgets/IWidget.cpp:147]: no {}

**Review by vaxerski — COMMENTED** [review]
logic-wise lgtm

**Review by vaxerski — APPROVED** [review]
thank you!

**Comment by michalrus** [comment]
Awesome, thanks! 💪 ✨

**Pull request merged**